graph/telemetry/istio/appender/istio_details.go

```diff
@@ -163,8 +163,18 @@ func addLabels(trafficMap graph.TrafficMap, globalInfo *graph.AppenderGlobalInfo
 	for _, n := range trafficMap {
 		// make sure service nodes have the defined app label so it can be used for app grouping in the UI.
 		if n.NodeType == graph.NodeTypeService && n.Namespace == sdl.Namespace.Name && n.App == "" {
-			// A service node that is a service entry will not have a service definition
+			// For service nodes that are a service entries, use the `hosts` property of the SE to find
+			// a matching Kubernetes Svc for adding missing labels
 			if _, ok := n.Metadata[graph.IsServiceEntry]; ok {
+				seInfo := n.Metadata[graph.IsServiceEntry].(*graph.SEInfo)
+				for _, host := range seInfo.Hosts {
+					if svc, found := svcMap[host]; found {
+						if app, ok := svc.Labels[appLabelName]; ok {
+							n.App = app
+						}
+						continue
+					}
+				}
```

Comment:
@israel-hdez (cc @nrfox)  I was originally thinking maybe the logic should be in the service_entry appender, when we create the service_entry node, but I guess this is an appropriate place as this is where we add the label for other service nodes.

The question I have is the same as what I asked in the issue, _Is it always true that the aggregated "se-service" nodes have the same "app" value, or is it possible to have "se-service" nodes with different apps, getting aggregated together?_.  Or, in other words, can a ServiceEntry node have multiple apps.  I don't really know the answer.  If we want to be very safe we could make sure only to make the assignment if there is one and only one app value.  But I'm also OK leaving this as-is.  I'll leave it up to you.

Reply:
You're asking if it's possible to have a ServiceEntry that looks something like:
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: details-svc
spec:
  hosts:
  - details.bookinfo.com
  - reviews.bookinfo.com
  location: MESH_INTERNAL
  ports:
  - number: 80
    name: http
    protocol: HTTP
  resolution: STATIC
  workloadSelector:
    labels:
      example: bookinfo
```
where one Service Entry has two hosts each pointing to a separate app? And how Kiali should aggregate the telemetry nodes in this case?

Reply:
I'm not necessarily questioning what we currently do to create the serviceEntry nodes.  But remember that a serviceEntry node can be an aggregate of multiple service nodes.  Because this PR is for assigning a single 'app' to the Se node, my question was whether there is actually at most one 'app' value to assign, and if not, how to handle.  So yes, in your example above, would you assign `details` or `reviews`.

Reply:
It seems like it's possible so maybe there is a valid use case out there for it.

If I'm understanding the change correctly, kiali is assuming that if a svc node without an app label has a service entry with a host that matches the name of a kube `Service`, then kiali sets the app label of the svc node to be the app label of the kube `Service` since the SE is "decorating" the kube `Service`?

What if there is no kube `Service` only the SE svc node? If the SE object has an "app: ..." label that matches the workload nodes, would it still get boxed with the workload nodes?

Reply:
> What if there is no kube `Service` only the SE svc node? If the SE object has an "app: ..." label that matches the workload nodes, would it still get boxed with the workload nodes?

It won't get boxed.
Anyway, yesterday I tried this case and I wasn't able to make the traffic to be routed properly. The bookinfo app always replied with something like "ratings is unavailable".

After repeatedly reading the docs ,what I understand is that, for the most part, Istio only deals with IPs. So, if you are trying to contact "reviews", you still need something to let your app to resolve "reviews" to an IP. Without the k8s svc, you will always get a "host not found" error. So, you need always a serviceentry AND a k8s service (or something else) to be able to resolve the DNS name.

Reply:
> Without the k8s svc, you will always get a "host not found" error.

When running inside of kube, you need to enable [dns proxying](https://istio.io/latest/docs/ops/configuration/traffic-management/dns-proxy/) first so that the proxy itself resolves the IP rather than kube dns. This is enabled by default on VM proxies.

Reply:
> When running inside of kube, you need to enable [dns proxying](https://istio.io/latest/docs/ops/configuration/traffic-management/dns-proxy/) first so that the proxy itself resolves the IP rather than kube dns. This is enabled by default on VM proxies.

I wasn't aware of it... Wow... apparently you sit and deeply studied the docs :-D

Reply:
> It would be nice for standalone (does not decorate an existing kube `Service`) SE nodes with an `app: ...` label to be boxed with their corresponding workload nodes.

Thinking a little bit, I think this is a completely different case IMO. I wasn't aware of the dns proxying feature, so I was unable to make it work. But, first, we would need to check what's the telemetry in that case. I'm not sure if Istio will record the name of the SE in `destination_service` labels. If not, currently, the graph may generate a workload-to-workload topology. If this is the case, we would need to inject the SE nodes and it's not that trivial to do, I think.

Reply:
> I'm not sure if Istio will record the name of the SE in `destination_service` labels. If not, currently, the graph may generate a workload-to-workload topology. If this is the case, we would need to inject the SE nodes and it's not that trivial to do, I think.

That's true it'll probably require some investigation and based on that it could greatly change the work required to implement. Perhaps that can be a separate issue from this one then and we can merge this as is? iirc from the bookinfo-vm demo, the service entry node does show up on the graph but it'd good be to verify that and get a better sense of what the telemetry is.

Reply:
> Perhaps that can be a separate issue from this one then and we can merge this as is?

+1 At least for the scope of the associated bug, I think this is good to go.
I can further investigate under another issue.